rust/relay/src/main.rs

```diff
@@ -19,36 +23,115 @@ async fn main() -> Result<()> {
-            let hex_bytes = hex::encode(&buf[..recv_len]);
-            tracing::trace!(target: "wire", r#"Input("{sender}","{}")"#, hex_bytes);
-        }
+                server.handle_relay_input(&payload, sender, allocation_id);
+            }
+            receive_result = pin!(socket.receive()) => {
+                let (sender, payload) = receive_result.context("Failed to receive from socket")?;
 
-        match server.handle_received_bytes(&buf[..recv_len], sender) {
-            Ok(Some((response, recipient))) => {
                 if tracing::enabled!(target: "wire", Level::TRACE) {
-                    let hex_bytes = hex::encode(&response);
-                    tracing::trace!(target: "wire", r#"Output("{recipient}","{}")"#, hex_bytes);
+                    let hex_bytes = hex::encode(&payload);
+                    tracing::trace!(target: "wire", r#"Input::Client("{sender}","{hex_bytes}")"#);
                 }
 
-                socket.send_to(&response, recipient).await?;
+                if let Err(e) = server.handle_client_input(&payload, sender, Instant::now()) {
+                    tracing::debug!("Failed to handle datagram from {sender}: {e}")
+                }
             }
-            Ok(None) => {}
-            Err(e) => {
-                tracing::debug!("Failed to handle datagram from {sender}: {e}")
+        }
+
+        while let Some(event) = server.next_command() {
+            match event {
+                Command::SendMessage { payload, recipient } => {
+                    if tracing::enabled!(target: "wire", Level::TRACE) {
+                        let hex_bytes = hex::encode(&payload);
+                        tracing::trace!(target: "wire", r#"Output::SendMessage("{recipient}","{hex_bytes}")"#);
+                    }
+
+                    socket.send_to(&payload, recipient).await?;
+                }
+                Command::AllocateAddresses { id, port } => {
+                    allocation_tasks.insert(id, tokio::spawn({
+                        let sender = relayed_data_sender.clone();
+
+                        async move {
+                            let Err(e) = forward_incoming_relay_data(sender, id, port).await else {
+                                unreachable!()
+                            };
+
+                            // TODO: Do we need to clean this up in the server? It will eventually timeout if not refreshed.
+                            tracing::warn!("Allocation task for {id} failed: {e}");
+                        }
+                    }));
+                }
+                Command::FreeAddresses { id } => {
+                    if let Some(task) = allocation_tasks.remove(&id) {
+                        tracing::info!("Freeing addresses of allocation {id}");
+                        task.abort();
+
+                        continue;
+                    }
+
+                    tracing::debug!("Unknown allocation {id}")
```

Comment:
[nit]!
```suggestion
                    } else {
                        tracing::debug!("Unknown allocation {id}") 
                    }
```

Reply:
That one was actually delibrate :)

I am not a fan of `else` because it puts both codepaths on the same indentation level, making it more difficult to visually distinguish.

Instead, I usually try to bail-early and have control-flow within the special-case. Anything non-indented is the happy path then.

Reply:
In writing this, I realised I flipped it here, will fix!